### PR TITLE
Add StateMachine.DeleteMemory

### LINKED ISSUE #58: HasMemory is broken
`StateMachine.HasMemory()` doesn't work as advertised. It's doing a check on a length of the byte buffer, which is not empty for a JSON marshaled string. It needs to be marshaled to a string before this check is done.

### PR DESCRIPTION
**Pull request by eviltoylet**
Fixes #58.

### CODE REVIEW

**Comment by njaremko** [comment]
To fix #58 , we should probably also include a check for an empty string in `HasMemory()`

**Comment by egtann** [comment]
@eviltoylet would love to merge this. Can you review the Contributor License Agreement, and sign it when you have an opportunity? It's based on the same CLA that Ubuntu/Canonical uses (called Harmony). Thanks!

**Comment by eviltoylet** [comment]
@egtann would love to sign the CLA, but I can't find it. Where's it located?

**Comment by egtann** [comment]
https://www.clahub.com/agreements/itsabot/abot

**Comment by eviltoylet** [comment]
@egtann signed. thanks!

File: shared/datatypes/state_machine.go [hunk 1]
```diff
@@ -280,6 +280,13 @@ func (sm *StateMachine) GetMemory(in *Msg, k string) Memory {
 	return Memory{Key: k, Val: buf, logger: sm.logger}
 }
 
+// DeleteMemory deletes a memory for a given key. It is not an error to delete a
+// key that does not exist.
+func (sm *StateMachine) DeleteMemory(in *Msg, k string) {
+	q := `DELETE FROM states WHERE userid=$1 AND pluginname=$2 AND key=$3`
+	sm.db.Exec(q, in.User.ID, sm.pluginName, k)
```
Comment (egtann) [inline shared/datatypes/state_machine.go]: To be consistent with the other Memory functions, we should be checking the value here and logging errors. See https://github.com/eviltoylet/abot/blob/master/shared/datatypes/state_machine.go#L260

Comment (eviltoylet) [inline shared/datatypes/state_machine.go]: I've added in the extra check. I'm not going to add the test in this checkin for the following reasons:
- I'm _really_ new to go and have no idea what the idiomatic way to test is
- there doesn't seem to be any existing infrastructure for a `DeleteMemory.` test to leverage. It's missing a sql mock driver.

**Comment by egtann** [comment]
Cool, no worries about the missing test. For what it's worth, we're using a real `abot_test` Postgres database, instead of a mock. It's easy to do and, because of that, makes it worthwhile to more closely mimic the production environment in our tests.

**Pull request merged**